src/mpi/init/init_async.c

```diff
@@ -59,22 +73,117 @@ static void progress_fn(void *data)
     return;
 }
 
+MPL_STATIC_INLINE_PREFIX int MPIDI_parse_progress_thread_affinity(int *thread_affinity,
+                                                                  int threads_per_node)
+{
+    int th_idx, read_count = 0, mpi_errno = MPI_SUCCESS;
+    char *affinity_copy = NULL;
+    const char *affinity_to_parse = MPIR_CVAR_CH4_PROGRESS_THREAD_AFFINITY;
+    char *proc_id_str, *tmp;
+    size_t proc_count;
+
+    if (!affinity_to_parse || strlen(affinity_to_parse) == 0) {
+        /* generate default affinity */
+        proc_count = MPL_get_nprocs();
+        for (th_idx = 0; th_idx < threads_per_node; th_idx++) {
+            if (th_idx < proc_count)
+                thread_affinity[th_idx] = proc_count - th_idx - 1;
+            else
+                thread_affinity[th_idx] = thread_affinity[th_idx % proc_count];
+        }
+        goto fn_exit;
+    }
+
+    /* create copy of original buffer because it will be modified in strsep */
+    affinity_copy = MPL_strdup(affinity_to_parse);
+    MPIR_Assert(affinity_copy);
+    tmp = affinity_copy;
+
+    for (th_idx = 0; th_idx < threads_per_node; th_idx++) {
+        proc_id_str = MPL_strsep(&tmp, " \n\t,");
+        if (proc_id_str != NULL) {
+            if (atoi(proc_id_str) < 0) {
+                printf
+                    ("ERROR: unexpected proc_id %s, affinity string %s\n",
+                     proc_id_str, affinity_to_parse);
+                mpi_errno = MPI_ERR_OTHER;
+                goto fn_fail;
+            }
+            thread_affinity[th_idx] = atoi(proc_id_str);
+            read_count++;
+        } else {
+            printf
+                ("ERROR: unexpected end of affinity string, expected %d numbers, read %d, affinity string %s\n",
+                 threads_per_node, read_count, affinity_to_parse);
+            mpi_errno = MPI_ERR_OTHER;
+            goto fn_fail;
+        }
+    }
+    if (read_count < threads_per_node) {
+        printf
+            ("ERROR: unexpected number of processors (specify 1 logical processor per 1 progress thread), affinity string %s\n",
+             affinity_to_parse);
+        mpi_errno = MPI_ERR_OTHER;
+        goto fn_fail;
+    }
+
+  fn_exit:
+    if (MPIR_Process.comm_world->rank == 0) {
+        for (th_idx = 0; th_idx < threads_per_node; th_idx++) {
+            MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                            (MPL_DBG_FDEST, "affinity: thread %d, processor %d",
+                             th_idx, thread_affinity[th_idx]));
+        }
```

Comment:
Move the debug logging to where we set the affinity, not where we parse the string.

Reply:
Moved.